File: CycleSort.py
def cycle_sort(arr):
    n = len(arr)
    for start in range(n):
        item = arr[start]
        pos = start

        for i in range(start + 1, n):
            if arr[i] < item:
                pos += 1

        if pos == start:
            continue

        while item == arr[pos]:
            pos += 1

        arr[pos], item = item, arr[pos]

        while pos != start:
            pos = start

            for i in range(start + 1, n):
                if arr[i] < item:
                    pos += 1

            while item == arr[pos]:
                pos += 1

            arr[pos], item = item, arr[pos]

if __name__ == "__main__":
    arr = [5, 2, 1, 8, 4]
    cycle_sort(arr)
    print("Sorted Array:", arr)

File: CycleSort.java
public class CycleSort {
    public static void cycleSort(int[] arr) {
        int n = arr.length;
        for (int start = 0; start < n - 1; start++) {
            int item = arr[start];
            int pos = start;

            for (int i = start + 1; i < n; i++) {
                if (arr[i] < item) {
                    pos++;
                }
            }

            if (pos == start) {
                continue;
            }

            while (item == arr[pos]) {
                pos++;
            }

            int temp = arr[pos];
            arr[pos] = item;
            item = temp;

            while (pos != start) {
                pos = start;
                for (int i = start + 1; i < n; i++) {
                    if (arr[i] < item) {
                        pos++;
                    }
                }

                while (item == arr[pos]) {
                    pos++;
                }

                temp = arr[pos];
                arr[pos] = item;
                item = temp;
            }
        }
    }

    public static void main(String[] args) {
        int[] arr = {5, 2, 1, 8, 4};
        cycleSort(arr);

        for (int num : arr) {
            System.out.print(num + " ");
        }
    }
}
